src/core/reader/mod.rs

```diff
@@ -4,73 +4,123 @@ use crate::{Error, Result};
+        }
+
+        self.pc = span.from;
+
+        Ok(())
     }
 
+    /// Byte slice to the remainder of the WASM binary, beginning from the current [pc](Self::pc)
     pub fn remaining_bytes(&self) -> &[u8] {
-        self.current
+        &self.full_wasm_binary[self.pc..]
     }
 
-    pub fn current_idx(&self) -> usize {
-        self.full_contents.len() - self.current.len()
-    }
-    pub fn make_span(&self, len: usize) -> Span {
-        Span::new(self.current_idx(), len)
+    /// Create a [Span] starting from [pc](Self::pc) for the next `len` bytes
+    ///
+    /// Does **not** verify the span to fit the WASM binary, i.e. the span could exceed the WASM
+    /// binarie's bounds.
+    pub fn make_span_unchecked(&self, len: usize) -> Span {
+        Span::new(self.pc, len)
     }
 
+    /// Take `N` bytes starting from [pc](Self::pc), then advance the [pc](Self::pc) by `N`
+    ///
+    /// This yields back an array of the correct length
     pub fn strip_bytes<const N: usize>(&mut self) -> Result<[u8; N]> {
-        if N > self.current.len() {
+        if N > self.full_wasm_binary.len() - self.pc {
             return Err(Error::Eof);
         }
 
-        let (bytes, rest) = self.current.split_at(N);
-        self.current = rest;
+        let bytes = &self.full_wasm_binary[self.pc..(self.pc + N)];
+        self.pc += N;
 
         Ok(bytes.try_into().expect("the slice length to be exactly N"))
     }
+
+    /// Read the current byte without advancing the [pc](Self::pc)
+    ///
+    /// May yield an error if the [pc](Self::pc) advanced past the end of the WASM binary slice
     pub fn peek_u8(&self) -> Result<u8> {
-        self.current.first().copied().ok_or(Error::Eof)
+        self.full_wasm_binary
+            .get(self.pc)
+            .copied()
+            .ok_or(Error::Eof)
     }
 
+    /// Call a closure that may mutate the [WasmReader]
+    ///
+    /// Returns a tuple of the closure's return value and the number of bytes that the WasmReader
+    /// was advanced by.
+    ///
+    /// # Panics
+    ///
+    /// May panic if the closure moved the [pc](Self::pc) backwards, e.g. when
+    /// [move_start_to](Self::move_start_to) is called.
     pub fn measure_num_read_bytes<T>(
         &mut self,
         f: impl FnOnce(&mut WasmReader) -> Result<T>,
     ) -> Result<(T, usize)> {
-        let before = self.current_idx();
+        let before = self.pc;
         let ret = f(self)?;
-        let num_read_bytes = self.current_idx() - before;
 
+        // TODO maybe use checked sub, that is slower but guarantees no surprises
+        debug_assert!(
+            self.pc >= before,
+            "pc was advanced backwards towards the start"
+        );
+
+        let num_read_bytes = self.pc - before;
         Ok((ret, num_read_bytes))
     }
 
+    /// Skip `num_bytes`, advancing the [pc](Self::pc) accordingly
+    ///
+    /// This can move the [pc](Self::pc) past the last byte of the WASM binary, so that reading
```

Comment:
The documentation is a bit contradictory here.
I'd suggest something like this instead:
```
If the resulting [pc](Self::pc) would end up past the last byte of the WASM binary,
an error is returned.
```

Reply:
This is also sort-of linked with the commend on `move_start_to`. `pc` can be set to any valid index within the slice of `wasm_bytes`, but also one past that.

|              Index |   0  |   1  |   2  |   3  |   4  | 5 | 6 |
|-------------------:|:----:|:----:|:----:|:----:|:----:|:-:|:-:|
|  wasm_bytes[index] | 0xaa | 0xbb | 0xcc | 0xee | 0xff | - | - |
| Valid `pc` position? |   ✅  |   ✅  |   ✅  |   ✅  |   ✅  | ✅ | ❌ |

The `pc` (if I understood correctly) means "the next position to read from", so, after consuming all valid wasm_bytes, the `pc` will end up at one index past the end. This comment says that the `pc` can **not** be moved by `skip` into an invalid position, but it can leave it in the "end" position, making any future reads panic.

Perhaps the wording on this is a bit vague, and "past the last byte in `wasm_bytes`" is a bit wordy. Perhaps we can explain this phenomenon in the documentation of the struct/field and then reference it in the function comments.

Reply:
I added a lot of documentation and more tests, and I copied your Table @valexandru . I hope this is now better documented.